components/brave_wallet/browser/brave_wallet_service.cc

```diff
@@ -692,4 +677,90 @@ void BraveWalletService::AddSignMessageRequest(
+  DCHECK(request->token->is_erc20 && !request->token->is_erc721);
+
+  if (add_suggest_token_requests_.contains(request->token->contract_address)) {
+    std::move(callback).Run(
+        false, mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_ALREADY_IN_PROGRESS_ERROR));
+    return;
+  }
+
+  const std::string addr = request->token->contract_address;
+  const std::string chain_id = GetCurrentChainId(prefs_);
+
+  // Priority of token source:
+  //   Mainnet:
+  //     1. User asset list
+  //     2. ERCTokenRegistry
+  //     3. wallet_watchAsset request
+  //   Others:
+  //     1. User asset list
+  //     2. wallet_watchAsset request
+  mojom::ERCTokenPtr token =
+      GetUserAsset(request->token->contract_address, request->token->token_id,
+                   request->token->is_erc721, chain_id);
+
+  if (!token && chain_id == mojom::kMainnetChainId)
+    token = ERCTokenRegistry::GetInstance()->GetTokenByContract(addr);
+
+  if (token)
+    request->token = std::move(token);
+
+  add_suggest_token_requests_[addr] = std::move(request);
+  add_suggest_token_callbacks_[addr] = std::move(callback);
+}
+
+void BraveWalletService::GetPendingAddSuggestTokenRequests(
+    GetPendingAddSuggestTokenRequestsCallback callback) {
+  std::vector<mojom::AddSuggestTokenRequestPtr> requests;
+  for (const auto& request : add_suggest_token_requests_) {
+    requests.push_back(request.second.Clone());
+  }
+  std::move(callback).Run(std::move(requests));
+}
+
+void BraveWalletService::NotifyAddSuggestTokenRequestsProcessed(
+    bool approved,
+    const std::vector<std::string>& contract_addresses) {
+  const std::string chain_id = GetCurrentChainId(prefs_);
+  for (const auto& addr : contract_addresses) {
+    if (add_suggest_token_requests_.contains(addr) &&
+        add_suggest_token_callbacks_.contains(addr)) {
+      auto callback = std::move(add_suggest_token_callbacks_[addr]);
+
+      if (approved &&
+          !AddUserAsset(add_suggest_token_requests_[addr]->token.Clone(),
+                        chain_id) &&
+          !SetUserAssetVisible(add_suggest_token_requests_[addr]->token.Clone(),
+                               chain_id, true)) {
+        add_suggest_token_requests_.erase(addr);
+        add_suggest_token_callbacks_.erase(addr);
+        std::move(callback).Run(
+            false, mojom::ProviderError::kInternalError,
+            l10n_util::GetStringUTF8(IDS_WALLET_INTERNAL_ERROR));
+        continue;
+      }
+
+      add_suggest_token_requests_.erase(addr);
+      add_suggest_token_callbacks_.erase(addr);
+      std::move(callback).Run(approved, mojom::ProviderError::kSuccess, "");
+    }
+  }
+}
+
+void BraveWalletService::OnNetworkChanged() {
+  add_suggest_token_requests_.clear();
+  // Reject pending suggest token requests when network changed.
+  for (auto& callback : add_suggest_token_callbacks_)
+    std::move(callback.second).Run(false, mojom::ProviderError::kSuccess, "");
```

Comment:
should it be `mojom::ProviderError::kUserRejectedRequest` here?

Reply:
https://github.com/brave/brave-core/pull/11518/commits/b795f4e65add1f1643a8f1e7aa91dd8120cd4dd6